### PR TITLE
Refactor docs generator HTML scaffold

### PR DESCRIPTION
**Pull request by straight-shoota**
This PR restructures some of the HTML templates for generating API docs. This is a piece extracted from #4755.

* prefixes partial templates with an underscore to distinguish them from full-page templates
* extracts repeating content of `<head>` tag to `_head.html` partial
* extracts sidebar to `_sidebar.html` partial
* refactors CSS to use flexbox instead of absolute positioning for the page layout

These are all just structural changes without changing any visuals. The only exception is that the sidebar now is always `30 em` wide instead of `20%` of screen size (which could get pretty tight on small screens).

### CODE REVIEW

**Comment by Sija** [comment]
Sidebar got bit wide after this change. Also, now it "jumps" after page change - to reproduce go to https://crystal-lang.org/api/master/, type search for some sth and hit enter, then on site refresh you'll see that sidebar goes expands in width and goes back to previous state after full site load.

**Comment by straight-shoota** [comment]
@Sija Hm, yeah that's because the width of the sidebar is influenced by the width of the main body and it adjust after that one is loaded/rendered. You can simulate this by toggling `display: none` on `.main-body` in the developer tools.

Now that this PR is merged, I plan to continue with a layout overhaul to make it adaptive for smaller screen sizes. This will be fixed then. I don't think it is worth a hot-fix.

File: src/compiler/crystal/tools/doc/html/css/style.css [hunk 1]
```diff
@@ -40,26 +40,28 @@ h2 {
...
+
+.sidebar, .main-content {
   overflow: auto;
 }
 
-#types-list {
-  left: 0;
-  width: 20%;
+.sidebar {
+  width: 30em;
```
Comment (RX14) [inline src/compiler/crystal/tools/doc/html/css/style.css]: This is unacceptable on vertical monitors.

Comment (straight-shoota) [reply src/compiler/crystal/tools/doc/html/css/style.css]: What exactly is unacceptable?

**Comment by bcardiff** [comment]
Do you think is a good idea to leave a downloadable version of the docs if the commit message matches `/\[doc(s?)\]/`? So from ci build anyone would be able to preview things a bit easier.

File: src/compiler/crystal/tools/doc/html/css/style.css (hunk 1 above)
Comment (RX14) [reply src/compiler/crystal/tools/doc/html/css/style.css]: That it takes up more than a third of my monitor.

**Comment by RX14** [comment]
@bcardiff why not every PR? We builds the docs in all CI runs anyway (iirc)

**Comment by bcardiff** [comment]
@RX14  I didn't want to store unnecessary artifacts. It seems that the quota is per file size and not per repo in circle...

https://circleci.com/docs/2.0/artifacts/
> Artifacts are stored on Amazon S3. There is a 3GB curl file size limit.

**Comment by straight-shoota** [comment]
I don't think having it downloadable is worth it. You can easily checkout the branch and run `make docs`. A real benefit would be having a preview to browse online.

File: src/compiler/crystal/tools/doc/html/css/style.css (hunk 1 above)
Comment (straight-shoota) [reply src/compiler/crystal/tools/doc/html/css/style.css]: Which browser and resolution?

Comment (Sija) [reply src/compiler/crystal/tools/doc/html/css/style.css]: True, on my MBA screen (1440×900) looks unnaturally wide too... :(

**Comment by Sija** [comment]
@straight-shoota 🏓

**Pull request merged**